lib_static/open_project/authentication/strategies/warden/doorkeeper_oauth.rb

```diff
@@ -26,25 +28,21 @@ def valid?
           def authenticate!
             access_token = ::Doorkeeper::OAuth::Token.authenticate(decorated_request,
                                                                    *Doorkeeper.configuration.access_token_methods)
-            return fail_with_header!(error: "invalid_token") if access_token.blank?
-            return fail_with_header!(error: "invalid_token") if access_token.expired? || access_token.revoked?
+            return fail_with_header!(error: "invalid_token") if access_token_invalid?(access_token)
             return fail_with_header!(error: "insufficient_scope") if !access_token.includes_scope?(scope)
 
-            if access_token.resource_owner_id.nil?
-              user_id = ::Doorkeeper::Application
-                          .where(id: access_token.application_id)
-                          .where.not(client_credentials_user_id: nil)
-                          .pick(:client_credentials_user_id)
-              authenticate_user(user_id) if user_id
-            else
-              authenticate_user(access_token.resource_owner_id)
-            end
+            user_id = access_token.resource_owner_id || access_token.application.client_credentials_user_id
+            authenticate_user(user_id) if user_id
```

Comment:
It seem there is no need to check for user_id presence, because we have a guard later in `authenticate_user(id)`
```suggestion
            authenticate_user(user_id)
```

Reply:
Haha, awesome. I was there before (you can see it hidden in my force push history).

There is a difference between the later guard and this one here: Here we just return `nil` on a missing user id, which means warden will ask the next strategy and fall through. In `authenticate_user` the guard will lead to an authentication error.

At first I changed the behaviour to be exactly like that, because looking at `spec/requests/api/v3/authentication_spec.rb` that was the better behaviour (it would return a proper WWW-Authenticate header there, not the Basic auth one).

However, then other specs would fail, because we specifically wanted to pass authentication through in that case. For certain configurations this allows to authenticate as anonymous user. I hope that my comments in the specs explain it well.

**tl;dr:** Removing the `if` would change behaviour compared to previous code, but the previous code was intentional (even if I don't 100% agree with the implementation). Keeping the if does not change existing behaviour, pure refactoring.

Reply:
True, you're right 👍